src/java/org/apache/cassandra/schema/Types.java

```diff
@@ -498,16 +570,101 @@ public long serializedSize(Types t, Version version)
             for (UserType type : t.types.values())
             {
                 size += sizeof(type.getNameAsString());
-                List<String> fieldNames = type.fieldNames().stream().map(FieldIdentifier::toString).collect(toList());
-                List<String> fieldTypes = type.fieldTypes().stream().map(AbstractType::asCQL3Type).map(CQL3Type::toString).collect(toList());
+                List<FieldIdentifier> fieldIdentifiers = type.fieldNames();
+                List<String> fieldNames = fieldIdentifiers.stream().map(FieldIdentifier::toString).collect(toList());
+                List<String> fieldTypes = type.fieldTypes().stream().map(AbstractType::asCQL3Type).map(CQL3Type::toString).collect(toList());;
                 size += sizeof(fieldNames.size());
                 for (String s : fieldNames)
                     size += sizeof(s);
                 size += sizeof(fieldTypes.size());
                 for (String s : fieldTypes)
                     size += sizeof(s);
+                if (version.isAtLeast(Version.V8))
+                {
+                    size += sizeof(type.comment);
+                    size += sizeof(type.securityLabel);
+                    size += fieldMetadataSize(type, fieldIdentifiers);
+                }
             }
             return size;
         }
+
+        private void serializeFieldMetadata(UserType type, List<FieldIdentifier> fieldIdentifiers, DataOutputPlus out) throws IOException
+        {
+            // Sparse serialization: only serialize non-empty metadata with their positions
+            serializeSparseMetadata(fieldIdentifiers, type::fieldComment, out);
+            serializeSparseMetadata(fieldIdentifiers, type::fieldSecurityLabel, out);
+        }
+
+        private void serializeSparseMetadata(List<FieldIdentifier> fieldIdentifiers,
+                                            java.util.function.Function<FieldIdentifier, String> metadataGetter,
+                                            DataOutputPlus out) throws IOException
+        {
+            // Count non-empty values
+            int nonEmptyCount = 0;
+            for (FieldIdentifier fieldId : fieldIdentifiers)
+            {
+                if (!metadataGetter.apply(fieldId).isEmpty())
+                    nonEmptyCount++;
+            }
+
+            // Write count followed by position-value pairs
+            out.writeInt(nonEmptyCount);
+            for (int i = 0; i < fieldIdentifiers.size(); i++)
+            {
+                String value = metadataGetter.apply(fieldIdentifiers.get(i));
+                if (!value.isEmpty())
+                {
+                    out.writeInt(i);
```

Comment:
hmm, I think it does not need to use `int` in practice. However, there is not restriction on limit the number of fields in a UDT. So theoretically, `int` is required. One can go really crazy and create Integer.MAX_VALUE number of fields. 
Add restriction is out of the scope of this PR. 
But I think you might want to write the variable int to avoid write bunch 0s.

```suggestion
                    out.writeUnsignedVInt32(i);
```